go/service/gregor_test.go

```diff
@@ -173,3 +182,206 @@ func TestShowTrackerPopupMessage(t *testing.T) {
 		t.Fatalf("Expected the tracker window for UID %s to be dismissed. current value: %s", trackee.User.GetUID().String(), identifyUI.dismissedUsername)
 	}
 }
+
+func newMsgID() gregor1.MsgID {
+	const mlen = 20
+	var letters = []byte("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	var res = make([]byte, mlen)
+	for i := 0; i < mlen; i++ {
+		res[i] = letters[rand.Intn(len(letters))]
+	}
+	return res
+}
+
+type mockGregord struct {
+	sm  gregor.StateMachine
+	fc  clockwork.FakeClock
+	log logger.Logger
+}
+
+func (m mockGregord) Sync(_ context.Context, arg gregor1.SyncArg) (gregor1.SyncResult, error) {
+	return grpc.Sync(m.sm, rpc.SimpleLogOutput{}, arg)
+}
+func (m mockGregord) ConsumeMessage(_ context.Context, msg gregor1.Message) error {
+	_, err := m.sm.ConsumeMessage(msg)
+	return err
+}
+func (m mockGregord) ConsumePublishMessage(_ context.Context, _ gregor1.Message) error {
+	return errors.New("unimplemented")
+}
+func (m mockGregord) Ping(_ context.Context) (string, error) {
+	return "pong", nil
+}
+
+func (m mockGregord) newIbm(uid gregor1.UID, msgid gregor1.MsgID) gregor1.Message {
+	m.fc.Advance(time.Minute)
+	return gregor1.Message{
+		Ibm_: &gregor1.InBandMessage{
+			StateUpdate_: &gregor1.StateUpdateMessage{
+				Md_: gregor1.Metadata{
+					Uid_:   uid,
+					MsgID_: msgid,
+					Ctime_: gregor1.ToTime(m.fc.Now()),
+				},
+				Creation_: &gregor1.Item{
+					Category_: "unknown!",
+					Body_:     gregor1.Body([]byte("HIHIHI")),
+				},
+			},
+		},
+	}
+}
+
+func newGregordMock(logger logger.Logger) mockGregord {
+	var of gregor1.ObjFactory
+	fc := clockwork.NewFakeClock()
+
+	sm := storage.NewMemEngine(of, fc)
+
+	return mockGregord{sm: sm, fc: fc, log: logger}
+}
+
+func setupSyncTests(t *testing.T, tc libkb.TestContext) (*gregorHandler, mockGregord, gregor1.UID) {
+	var err error
+	user, err := kbtest.CreateAndSignupFakeUser("gregr", tc.G)
+	if err != nil {
+		t.Fatal(err)
+	}
+	uid := gregor1.UID(user.User.GetUID().ToBytes())
+
+	var h *gregorHandler
+	if h, err = newGregorHandler(tc.G); err != nil {
+		t.Fatal(err)
+	}
+
+	server := newGregordMock(tc.G.Log)
+
+	return h, server, uid
+}
+
+func checkStateSize(t *testing.T, message string, h *gregorHandler, size int) {
+	var state gregor.State
+	var err error
+	if state, err = h.gregorCli.StateMachineState(nil); err != nil {
+		t.Fatal(err)
+	}
+
+	items, err := state.Items()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(items) != size {
+		t.Fatalf(message, len(items), size)
+	}
+}
+
+func TestSyncBasic(t *testing.T) {
+	tc := libkb.SetupTest(t, "gregor")
+	defer tc.Cleanup()
+	tc.G.SetService()
+
+	// Set up client and server
+	h, server, uid := setupSyncTests(t, tc)
+
+	//Consume a bunch of messages to the server, and we'll sync them down
```

Comment:
The next test is supposed to check to see if we properly only sync down the messages we need, this first one is just that the basic functionality works.